### PR TITLE
bpo-33407: Implement Py_DEPRECATED() on Windows

### PR DESCRIPTION
**Pull request by ZackerySpytz**
https://bugs.python.org/issue33407

### CODE REVIEW

**Review by vstinner — APPROVED** [review]
LGTM.

@serhiy-storchaka, @zooba: Would you mind to double check?

**Comment by serhiy-storchaka** [comment]
The current usage:
```c
extern int foo(int) Py_DEPRECATED(3.2);
```
This PR changes it to
```c
Py_DEPRECATED(3.2) extern int foo(int);
```
It requires to rewrite *all* code that uses `Py_DEPRECATED`. This is a large breakage. `Py_DEPRECATED` is public and can be used in user code. We shouldn't break it in such way.

**Comment by vstinner** [comment]
Ok, but we have to make the change to support Visual Studio. Do you think
that the macro is used outside Python itself?

**Review by vstinner — CHANGES_REQUESTED** [review]
@ZackerySpytz: Can you please explain that Py_DEPRECATED() macro must now be used at the start of the function definition, rather than at the end, in a new "Changes in the C API" section at https://docs.python.org/dev/whatsnew/3.8.html#porting-to-python-3-8 ?

See https://docs.python.org/dev/whatsnew/3.7.html#changes-in-the-c-api for an example of the Python 3.7 documentation.

@ZackerySpytz: Can you please confirm that using Py_DEPRECATED() at the start of the function (rather than the end) already works in Python 3.7?

**Comment by ZackerySpytz** [comment]
@vstinner I have added a What's New entry for this change.
> Can you please confirm that using Py_DEPRECATED() at the start of the function (rather than the end) already works in Python 3.7?

Yes, I can confirm using Py_DEPRECATED() at the start of the function already worked before this change.
I have made the requested changes; please review again.

File: Doc/whatsnew/3.8.rst [hunk 1]
```diff
@@ -1077,6 +1077,15 @@ Changes in the C API
 
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The :c:macro:`Py_DEPRECATED()` macro has been implemented for MSVC.
+  The macro now must be placed before the symbol name.
+
+  Example::
+
+      Py_DEPRECATED(3.8) int Py_OldFunction(void);
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.8.rst]: Should not we use `PyAPI_FUNC()`?

File: Include/pyport.h [hunk 2]
```diff
@@ -505,13 +505,16 @@ extern "C" {
 /* Py_DEPRECATED(version)
  * Declare a variable, type, or function deprecated.
  * Usage:
- *    extern int old_var Py_DEPRECATED(2.3);
- *    typedef int T1 Py_DEPRECATED(2.4);
- *    extern int x() Py_DEPRECATED(2.5);
+ *    Py_DEPRECATED(2.3) extern int old_var;
```
Comment (serhiy-storchaka) [inline Include/pyport.h]: Did you test this?

Comment (ZackerySpytz) [reply Include/pyport.h]: Yes.

File: Doc/whatsnew/3.8.rst (hunk 1 above)
Comment (ZackerySpytz) [reply Doc/whatsnew/3.8.rst]: Sure, but this is just an example, no?

Comment (serhiy-storchaka) [reply Doc/whatsnew/3.8.rst]: Why not use a good example? It seems to me, that `PyAPI_FUNC()` is necessary in all cases.

Comment (ZackerySpytz) [reply Doc/whatsnew/3.8.rst]: Okay, I will add `PyAPI_FUNC()`.

File: Include/pyport.h (hunk 2 above)
Comment (ZackerySpytz) [reply Include/pyport.h]: Sorry, by "yes" I meant that this example compiles, but it is actually a useless example and should be removed.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by ZackerySpytz** [comment]
@vstinner, I have implemented all of your requested changes, and Serhiy Storchaka has now approved this PR. Please have a look.

File: Misc/NEWS.d/next/Windows/2018-08-28-17-23-49.bpo-33407.ARG0W_.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Implement ``Py_DEPRECATED()`` on Windows.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Windows/2018-08-28-17-23-49.bpo-33407.ARG0W_.rst]: Windows != MSVC. There are clang, Cygwin, MinGW, etc. C compilers on Windows. Please be more specific:

```suggestion
Implement ``Py_DEPRECATED()`` on MSVC.
```

Comment (vstinner) [reply Misc/NEWS.d/next/Windows/2018-08-28-17-23-49.bpo-33407.ARG0W_.rst]: Or maybe just reuse your sentence from whatsnew:

```
The :c:macro:`Py_DEPRECATED()` macro has been implemented for MSVC. 
```

File: Doc/whatsnew/3.8.rst [hunk 4]
```diff
@@ -1077,6 +1077,15 @@ Changes in the C API
 
   (Contributed by Eddie Elizondo in :issue:`35810`.)
 
+* The :c:macro:`Py_DEPRECATED()` macro has been implemented for MSVC.
+  The macro now must be placed before the symbol name.
+
+  Example::
+
+      Py_DEPRECATED(3.8) PyAPI_FUNC(int) Py_OldFunction(void);
```
Comment (vstinner) [inline Doc/whatsnew/3.8.rst:1085]: Would you mind to document the macro at:
https://docs.python.org/dev/c-api/intro.html#useful-macros

Basically, just copy/paste your text there.

File: Include/pyport.h [hunk 5]
```diff
@@ -505,13 +505,16 @@ extern "C" {
...
+ *    Py_DEPRECATED(2.4) typedef int T1;
+ *    Py_DEPRECATED(2.5) extern int x();
```
Comment (vstinner) [inline Include/pyport.h]: Please document usage for functions:

```
      Py_DEPRECATED(3.8) PyAPI_FUNC(int) Py_OldFunction(void);
```

niptick: would you mind to upgrade the Python version to 3.x? :-D Just replace 2 with 3.

File: Include/pyport.h [hunk 6]
```diff
@@ -505,13 +505,16 @@ extern "C" {
 /* Py_DEPRECATED(version)
  * Declare a variable, type, or function deprecated.
```
Comment (vstinner) [inline Include/pyport.h:506]: Please add: "The macro must be placed before the symbol name."

**Review by vstinner — CHANGES_REQUESTED** [review]
I have requests on the documentation part.

File: Include/pyport.h (hunk 6 above)
Comment (serhiy-storchaka) [reply Include/pyport.h:506]: Before the declaration.

Comment (ZackerySpytz) [reply Include/pyport.h:506]: Done.

File: Include/pyport.h (hunk 5 above)
Comment (ZackerySpytz) [reply Include/pyport.h]: Done.

File: Misc/NEWS.d/next/Windows/2018-08-28-17-23-49.bpo-33407.ARG0W_.rst (hunk 3 above)
Comment (ZackerySpytz) [reply Misc/NEWS.d/next/Windows/2018-08-28-17-23-49.bpo-33407.ARG0W_.rst]: I know that there are a variety of C compilers usable on Windows, but as MSVC is currently the only one that is supported with CPython on Windows, I thought it was okay to just use "Windows" in the new entry.

**Comment by ZackerySpytz** [comment]
Two deprecation warnings are emitted on MSVC because of this PR; they can be seen on AppVeyor (e.g. https://ci.appveyor.com/project/python/cpython/builds/24580662#L125):

```
_winapi.c
c:\projects\cpython\modules\_winapi.c(505): warning C4996: 'PyErr_SetFromWindowsErrWithUnicodeFilename': deprecated in 3.3 [C:\projects\cpython\PCbuild\pythoncore.vcxproj]
  c:\projects\cpython\include\cpython\pyerrors.h(108): note: see declaration of 'PyErr_SetFromWindowsErrWithUnicodeFilename'
c:\projects\cpython\modules\_winapi.c(1382): warning C4996: 'PyErr_SetFromWindowsErrWithUnicodeFilename': deprecated in 3.3 [C:\projects\cpython\PCbuild\pythoncore.vcxproj]
  c:\projects\cpython\include\cpython\pyerrors.h(108): note: see declaration of 'PyErr_SetFromWindowsErrWithUnicodeFilename'
```

I had opened https://bugs.python.org/issue36935 and GH-13355 to fix these warnings.

**Comment by ZackerySpytz** [comment]
I have made the requested changes; please review again.

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**